libminifi/src/io/ZlibStream.cpp

```diff
@@ -0,0 +1,202 @@
+  strm_.zalloc = Z_NULL;
+  strm_.zfree = Z_NULL;
+  strm_.opaque = Z_NULL;
+}
+
+bool ZlibBaseStream::isFinished() const {
+  return state_ == ZlibStreamState::FINISHED;
+}
+
+/* ZlibCompressStream */
+
+ZlibCompressStream::ZlibCompressStream(bool gzip, int level)
+  : ZlibCompressStream(this, gzip, level) {
+}
+
+ZlibCompressStream::ZlibCompressStream(DataStream* other, bool gzip, int level)
+  : ZlibBaseStream(other)
+  , logger_(logging::LoggerFactory<ZlibCompressStream>::getLogger()) {
+  int ret = deflateInit2(
+      &strm_,
+      level,
+      Z_DEFLATED /* method */,
+      15 + (gzip ? 16 : 0) /* windowBits */,
+      8 /* memLevel */,
+      Z_DEFAULT_STRATEGY /* strategy */);
+  if (ret != Z_OK) {
+    logger_->log_error("Failed to initialize z_stream with deflateInit2, error code: %d", ret);
+    throw Exception(ExceptionType::GENERAL_EXCEPTION, "zlib deflateInit2 failed");
+  }
+
+  state_ = ZlibStreamState::INITIALIZED;
+}
+
+ZlibCompressStream::~ZlibCompressStream() {
+  if (state_ != ZlibStreamState::UNINITIALIZED) {
+    deflateEnd(&strm_);
+  }
+}
+
+int ZlibCompressStream::writeData(uint8_t* value, int size) {
+  if (state_ != ZlibStreamState::INITIALIZED) {
+    logger_->log_error("writeData called in invalid ZlibCompressStream state, state is %hhu", state_);
+    return -1;
+  }
+
+  strm_.next_in = value;
+  strm_.avail_in = size;
+
+  /*
+   * deflate consumes all input data it can (i.e. if it has enough output buffer it never leaves input data unconsumed)
+   * and fills the output buffer to the brim every time it can. This means that the proper way to use deflate is to
+   * watch avail_out: once deflate does not have to fill it to the brim, then it has consumed all data we have provided
+   * to it, and does not need more output buffer for the time being.
+   * When we have no more input data, we must call deflate with Z_FINISH to make it empty its internal buffers and
+   * close the compressed stream.
+   */
+  do {
+    logger_->log_trace("writeData has %u B of input data left", strm_.avail_in);
+
+    strm_.next_out = outputBuffer_.data();
+    strm_.avail_out = outputBuffer_.size();
+
+    int flush = value == nullptr ? Z_FINISH : Z_NO_FLUSH;
+    logger_->log_trace("calling deflate with flush %d", flush);
+
+    int ret = deflate(&strm_, flush);
+    if (ret == Z_STREAM_ERROR) {
+      logger_->log_error("deflate failed, error code: %d", ret);
+      state_ = ZlibStreamState::ERRORED;
+      return -1;
+    }
+    int output_size = outputBuffer_.size() - strm_.avail_out;
+    logger_->log_trace("deflate produced %d B of output data", output_size);
+    if (BaseStream::writeData(outputBuffer_.data(), output_size) != output_size) {
+      logger_->log_error("Failed to write to underlying stream");
+      state_ = ZlibStreamState::ERRORED;
+      return -1;
+    }
+  } while (strm_.avail_out == 0);
+
+  return size;
+}
+
+void ZlibCompressStream::closeStream() {
+  if (state_ == ZlibStreamState::INITIALIZED) {
```

Comment:
In case error happened during write and close is called, what happens? Or we don't expect it to happen at all?

Reply:
It's perfectly okay to call `closeStream` if an error happens, and we will do nothing: we will keep the error state as it was.